javalib/src/main/scala/java/io/CharArrayWriter.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.io
+
+class CharArrayWriter(initialSize: Int)  extends Writer {
+  if (initialSize < 0)
+    throw new IllegalArgumentException("size must be >= 0")
+
+  protected var buf: Array[Char] = new Array[Char](initialSize)
+  protected var count: Int = 0
+
+  def this() = this(32)
+
+  override def close(): Unit = ()
+
+  private def ensureCapacity(i: Int): Unit = {
+    if (this.count + i > this.buf.length) {
+      val newLen = Math.max(2 * this.buf.length, this.count + i)
```

Comment:
You are referring to the int overflow, right?  Can you explain what the potential issue is here to me like I'm a junior dev?  It's still non obvious to me what is being protected here, and want to make sure I understand.

Reply:
For the `if` line:

- Assume `this.count = Int.MaxValue - 1`
- Call `ensureCapacity(2)`
- `this.count + i == Int.MinValue`
- `Int.Min.Value < this.buf.length`
- Therefore, `ensureCapacity` does nothing, writing will overflow.

For the line setting `newLen`:
If `this.buf.length = Int.MaxValue / 2 + 1`, then `2 * this.buf.length` is negative. As long a `this.count + i` is not negative, that's not a problem. But otherwise, `newLen` would be negative. This would make `copyOf` fail.